concordium-std/src/state_btree.rs

```diff
@@ -0,0 +1,1481 @@
+        }
+        // Transmuting away the maybeuninit is safe since we have initialized all of
+        // them.
+        unsafe { mem::transmute(prefixed) }
+    }
+}
+
+impl<const M: usize, K, V, S> Serial for StateBTreeMap<K, V, S, M> {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        self.map.serial(out)?;
+        self.ordered_set.serial(out)
+    }
+}
+
+impl<const M: usize, K, V, S: HasStateApi> DeserialWithState<S> for StateBTreeMap<K, V, S, M> {
+    fn deserial_with_state<R: Read>(state: &S, source: &mut R) -> ParseResult<Self> {
+        let map = DeserialWithState::deserial_with_state(state, source)?;
+        let ordered_set = DeserialWithState::deserial_with_state(state, source)?;
+        Ok(Self {
+            map,
+            ordered_set,
+        })
+    }
+}
+
+impl<const M: usize, K, S> Serial for StateBTreeSet<K, S, M> {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        self.prefix.serial(out)?;
+        self.root.serial(out)?;
+        self.len.serial(out)?;
+        self.next_node_id.serial(out)
+    }
+}
+
+impl<const M: usize, K, S: HasStateApi> DeserialWithState<S> for StateBTreeSet<K, S, M> {
+    fn deserial_with_state<R: Read>(state: &S, source: &mut R) -> ParseResult<Self> {
+        let prefix = source.get()?;
+        let root = source.get()?;
+        let len = source.get()?;
+        let next_node_id = source.get()?;
+
+        Ok(Self {
+            _marker_key: Default::default(),
+            prefix,
+            state_api: state.clone(),
+            root,
+            len,
+            next_node_id,
+        })
+    }
+}
+
+impl<const M: usize, K: Serial> Serial for Node<M, K> {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        self.keys.serial(out)?;
+        self.children.serial(out)
+    }
+}
+
+impl<const M: usize, K: Deserial> Deserial for Node<M, K> {
+    fn deserial<R: Read>(source: &mut R) -> ParseResult<Self> {
+        let keys = source.get()?;
+        let children = source.get()?;
+        Ok(Self {
+            keys,
+            children,
+        })
+    }
+}
+
+impl Serial for NodeId {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> { self.id.serial(out) }
+}
+
+impl Deserial for NodeId {
+    fn deserial<R: Read>(source: &mut R) -> ParseResult<Self> {
+        Ok(Self {
+            id: source.get()?,
+        })
+    }
+}
+
+impl<const M: usize, K, V, S> Deletable for StateBTreeMap<K, V, S, M>
+where
+    S: HasStateApi,
+    K: Serialize,
+    V: Serial + DeserialWithState<S> + Deletable,
+{
+    fn delete(mut self) { self.clear(); }
+}
+
+/// This test module rely on the runtime providing host functions and can only
+/// be run using `cargo concordium test`.
+#[cfg(feature = "wasm-test")]
+mod wasm_test_btree {
+    use crate::{claim, claim_eq, concordium_test, StateApi, StateBuilder};
+
+    #[concordium_test]
+    fn test_btree_insert_6() {
```

Comment:
Could you please have a sentence or two for each of these tests to explain what it's testing?